docs/rfcs/021-raii-runtime.md

```diff
@@ -0,0 +1,209 @@
+# Improving runtime by making use of the RAII idiom
+
+## Motivation
+
+Currently, we have several problems regarding the runtime:
+
+- `TenantState` doesn't clearly show what happens with the tenant. Repository and code comments need to describe how state transitions occur and what they mean. It is critical since the code is going to have more states. The same applies to `TimelineState`.
+- Current tenant and timeline access, shutdown design fails to shut down tenants properly, timelines, background tasks, and control their lifetimes, resulting in races.
+- `tasks_mgr` runtime is redundant, and its design allows us to have even more races without noticing.
+
+This RFC describes how to rewrite a runtime in a more RAII fashion, fixing part of its problems.
+
+## Small quality of life changes
+
+Before we start, some synchronisation primitives should be proposed.
+
+1. In our code, there're multiple places where we use `tokio::watch::{Sender, Receiver}` with type `()`. It's hard to track down whether we're just waiting for the drop, for a signal or both from the sender. To make it clear, a couple of wrappers are needed:
+    - `SignalSender` / `SignalWatcher`: wait only for proper signals and `panic!` when it receives `RecvError`.
+    - `DropSender` / `DropWatcher`: wait only for `RecvError` and don't have any function in the interface to signal.
+    - Non-generic `Sender` / `Receiver` wrappers for both drops and signals.
+2. For the cancellation proposal, we'll need an object that both can wait for a task to finish and send a cancellation request. We'll call it `CancelationSender` / `CancellationReceiver`. It could be coded using two tokio channels.
+3. For the cancellation proposal, we'll need a triple `StateKeeper` / `StateSubscriber` / `StateHider`. This triple atomically changes the state and sends it to subscribers like `tokio::watch`. But when there's at least one `StateHider`, we put the new state to the buffer that is being read only by subscribers, and `StateKeeper` will read the older value. After all `StateHider`s are gone, we atomically (with respect to the new possible `StateHider`s) change the state.
```

Comment:
I don't think we need the `StateKeeper` and `StateHider`. If I understand correctly, the idea is that if you start an operation when the tenant is in Active state, for example, and the state is changed concurrently to ShuttingDown, the operation that started earlier sees the state still as Active. Kind of like MVCC: you see the state as it was when the operation started. I don't think that's necessary. If an operation starts when the tenant is Active, it has no need to look at the old state. It probably won't have any reason to look at the state at all, as long as the accessor/guard system ensures that the tenant stays in Active or ShuttingDown state until the guard is dropped.

Reply:
I didn't know that's called MVCC :)

I think you're right. I removed that from RFC. From this moment, closures will have a guarantee they started when the task was in the `Active` state, and after it becomes `ShuttingDown`, we won't have new closures starts.